### PR TITLE
Updated templates and materials

### PR DESCRIPTION
**Pull request by Rackover**
Fixed normal path (black stripes) and remade every biome with the template generator.
Also added 4 new biomes.
Here are all the biomes and proof that they look correct ingame.

Debris 
Debris

Glacier
Glacier

Haven
Haven

Noon
Noon

Overlord
Overlord

Refuge
Haven

Shuriken
Shuriken

Syrtis
Syrtis

Wonder
Wonder

### CODE REVIEW

**Comment by Rackover** [comment]
New preview generation :

File: src/java/export/SCMapExporter.java [hunk 1]
```diff
@@ -172,6 +175,16 @@ public static void exportSCMAP(Path folderPath, String mapname, SCMap map) throw
...
 		out.close();
+
+		String fileFormat = "png";
+		File previewFile = folderPath.resolve(mapname).resolve(mapname + "_preview." + fileFormat).toFile();
+		RenderedImage renderedImage = map.getPreview();
+		try{
+			ImageIO.write(renderedImage, fileFormat, previewFile);
+		}
+		catch (IOException e) {
+			System.out.println("Could not write the preview image\n"+e.toString());
```
Comment (Geosearchef) [inline src/java/export/SCMapExporter.java]: printf

File: src/java/export/SCMapExporter.java [hunk 2]
```diff
@@ -172,6 +175,16 @@ public static void exportSCMAP(Path folderPath, String mapname, SCMap map) throw
 
 		out.flush();
 		out.close();
+
+		String fileFormat = "png";
```
Comment (Geosearchef) [inline src/java/export/SCMapExporter.java]: constant

File: src/java/generator/MapGenerator.java [hunk 3]
```diff
@@ -206,39 +210,32 @@ public SCMap generate(long seed) throws ExecutionException, InterruptedException
 		map.setTextureMaskLow(grassTexture.getFloatMask(), lightGrassTexture, rockTexture.getFloatMask(), new FloatMask(513, 0));
 
 		Biome biomeSet = Biomes.list.get(random.nextInt(Biomes.list.size()));
+		System.out.println(String.format("Using biome %s", biomeSet.getName()));
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: printf

File: src/java/generator/MapGenerator.java [hunk 4]
```diff
@@ -206,39 +210,32 @@ public SCMap generate(long seed) throws ExecutionException, InterruptedException
...
-					blue = 96;
-				}
-				g.setColor(new Color(red, green, blue));
+				int scaledX = x*2;
+				int scaledY = y*2;
+
+				float delta = ((float)map.getHeightmap().getRaster().getSample(scaledX,scaledY,0))
+						* SCMap.HEIGHTMAP_SCALE
+						* (1/50f); // Magic number to make the colors more readable
+				delta = (Math.round(delta*heatmapGradientSteps + 0.5f) -0.5f)/heatmapGradientSteps;
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: StrictMath.round()

File: src/java/generator/MapGenerator.java [hunk 5]
```diff
@@ -206,39 +210,32 @@ public SCMap generate(long seed) throws ExecutionException, InterruptedException
...
+				delta = (1f-heatmapDeadZone) - delta*(1f-heatmapDeadZone*2f);
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: space between operator and operands pls

File: src/java/map/Material.java [hunk 6]
```diff
@@ -22,8 +23,8 @@ public Material(String environment, String texture, String normal, float texture
         this(environment, environment, texture, normal, textureScale, normalScale);
     }
     public Material(String texEnv, String normalEnv, String texture, String normal, float textureScale, float normalScale) {
-        texturePath = String.format(pathFormat, texEnv, texture, "albedo");
-        normalPath = String.format(pathFormat, normalEnv, normal, "normals");
+        texturePath = String.format(texturePathFormat, texEnv, texture, "albedo");
+        normalPath = String.format(normalPathFormat, normalEnv, normal);
```
Comment (Geosearchef) [inline src/java/map/Material.java:27]: @n3roxis pls verify this

**Review by Geosearchef — CHANGES_REQUESTED** [review]
Please seperate the image generation into another pull request

File: src/java/map/Material.java (hunk 6 above)
Comment (louvenarde) [reply src/java/map/Material.java:27]: verify what ?
i've updated the template generator that I wrote myself aswell as this part of the code that used the same layout as the template generator and updated both (aswell as all the templates) according to it
Look at the screenshots in the PR
and look at the JSON files aswell

**Comment by Rackover** [comment]
Better version

File: src/java/export/SCMapExporter.java [hunk 7]
```diff
@@ -172,6 +175,16 @@ public static void exportSCMAP(Path folderPath, String mapname, SCMap map) throw
...
 		out.close();
+
+		final String fileFormat = "png";
+		File previewFile = folderPath.resolve(mapname).resolve(mapname + "_preview." + fileFormat).toFile();
+		RenderedImage renderedImage = map.getPreview();
+		try{
+			ImageIO.write(renderedImage, fileFormat, previewFile);
+		}
+		catch (IOException e) {
+			System.out.printf("Could not write the preview image\n"+e.toString());
```
Comment (Geosearchef) [inline src/java/export/SCMapExporter.java:185]: printf

File: src/java/generator/MapGenerator.java [hunk 8]
```diff
@@ -206,42 +210,22 @@ public SCMap generate(long seed) throws ExecutionException, InterruptedException
...
-				if (rock.getBinaryMask().get(x, y)) {
-					red = 96;
-					green = 96;
-					blue = 96;
-				}
-				g.setColor(new Color(red, green, blue));
-				g.fillRect(x, y, 1, 1);
-			}
-		}
+	    Preview.generate(
```
Comment (Geosearchef) [inline src/java/generator/MapGenerator.java]: indent

File: src/java/generator/Preview.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+                float elevation =
+                        ((float)map.getHeightmap().getRaster().getSample(scaledX,scaledY,0))
+                                * SCMap.HEIGHTMAP_SCALE;
+
+                float delta = elevation * (1/60f); // Magic number to make the colors more readable
+
+                Color color = HEATMAP_GRADIENT.evaluate(delta);
+                float[] hsb = Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), null);
+                Color desaturated = Color.getHSBColor(hsb[0], hsb[1]*HEATMAP_SATURATION, hsb[2]*HEATMAP_GAMMA);
```
Comment (Geosearchef) [inline src/java/generator/Preview.java]: spacing pls

File: src/java/map/SCMap.java [hunk 10]
```diff
@@ -6,6 +6,7 @@
 import util.Vector4f;
 import util.serialized.WaterSettings;
 
+import java.awt.*;
```
Comment (Geosearchef) [inline src/java/map/SCMap.java]: pls run optimize imports over the entire project (select src folder + press ctrl+alt+o)

File: src/java/util/Gradient.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import lombok.AllArgsConstructor;
+
+import java.awt.*;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+public strictfp class Gradient {
+    public Map<Float, Color> colors = new TreeMap<>();
```
Comment (Geosearchef) [inline src/java/util/Gradient.java]: private

File: src/java/util/Gradient.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public Color evaluate(float period){
+        if (period > 1f || period < 0f){
+            throw new RuntimeException("Period must be comprised between 0 and 1 (supplied: "+period+")");
+        }
+
+        float previousKey = 0f;
+        float nextKey = 1f;
+        for (float key : colors.keySet()){
```
Comment (Geosearchef) [inline src/java/util/Gradient.java]: you're using a red black tree for storing the map https://en.wikipedia.org/wiki/Red%E2%80%93black_tree . You can use this data structure to search for an index minimally greater or lower than the supplied one, see: https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#higherEntry-K-

**Pull request merged**